### DISCUSSION TITLE
Branching Strategy

### DISCUSSION THREAD

**Discussion by andregasser** [andregasser/bigbone#45] [Ideas]
I was thinking about a potential branching strategy for this library, given the situation, that there will be multiple releases pushed to the artifact repository (e.g. Maven Central) at the same time. Applying a branching strategy as described below would give us the freedom to implement bugfixes on earlier releases, without affecting the current development on the next version.

https://www.bitsnbites.eu/a-stable-mainline-branching-model-for-git/

Any feedback on this?

**Comment by bocops** [comment]
Just seeing this now, seems like a generally good strategy. One thing I wonder about, though, is how this strategy would work with both breaking and non-breaking changes in the master branch. An example:

- master branch commit A becomes 2.0.0-rc1 in its own release branch
- later, commit B is added to the master branch. This is a breaking change that we would publish in a 3.0.0 release but not before.
- later, commit C is added to the master branch. This is a non-breaking change that could be published in a 2.1.0 release.

How would we release the changes of commit C, if it has breaking changes (commit B) in its commit history? The first image in the linked article claims that we could branch 2.1.0 off of master at some point - but then we would need to make sure that no non-breaking change gets committed at any point before.

The alternative would be to commit non-breaking changes both to the main and (major) release branch, but then we'd run into the same problems with non-breaking vs. bug-fixing changes, where one would increase the _minor_ and the other the _fix_ part of the version number.

For real-world examples of breaking vs. non-breaking changes, look at issues:
- #24 (incremental change that could be released as 2.1.0)
- #62 (breaking change, unless we keep the old API until the next major release after adding the new one)

**Comment by andregasser** [comment]
Thanks for your input on this. Eventually, I was thinking too big in this case (as often 🙂). To be honest, I am not really eager on maintaining a complex release branching strategy for this library. Maybe we should keep this really simple and make sure, we can focus on feature development instead of wasting time with merging / cherry picking commits between branches all the time. 

I have read about "trunk-based development" a bit (which seems to be the new standard for fast moving teams). It has some interesting approaches in it. I like it because it is simple. Here's an article about it: (there might be better ones): https://www.atlassian.com/continuous-delivery/continuous-integration/trunk-based-development

Based on the article above, I would like to come up with a new approach:

- Our library follows semantic versioning practices (as already agreed earlier)
- Our repository holds a `master` branch and a number of feature branches
- Our `master` branch is always in a releasable state (means it is green, all tests passing)
- Test coverage is high and gives confidence that our code in the `master` is good enough to be released
- We release often (on demand, even once a week)
- Features and fixes of any kind (e.g. security issues) will be released as a new version. We do not maintain older version. Example:
  - Security vulnerability in version 2.1.1 --> we release a version 2.1.2 that fixes it
  - New non-breaking changes added to 2.1.1 --> We release a 2.2.0 version
  - New breaking changes added to 2.1.1 --> We release a 3.0.0 version

This has some advantages for us:
- We get new features out to developers quicker
- We can fix issues fast if needed (as we can release anytime we want)

But it has also advantages for library users:
- They get their features quicker
- Changelogs are shorter, that makes upgrading to the next version easier I guess. 
- Also, given the fact that we do smaller releases, they can gradually upgrade to the latest version, and make sure, their code works. Example: Developer is on version 2.1.1, we have released versions 2.2.0, 2.2.1, 2.3.0, 3.0.0 and 3.1.0 after. So he can start to upgrade his project to 2.20, then to 2.2.1 up to 3.1.0 in a step-by-step fashion.

Also, if we need to introduce a breaking change, given current version 2.2.3, we can just release a new version 3.0.0. The change log makes it clear what the breaking change is. But as the release 3.0.0 itself is a small release, adopting to 3.0.0 should not be too cumbersome for the dev.

I think it is acceptable to have a policy like this. What do you say?

**Comment by bocops** [comment]
I assume that we will have many breaking changes in the near future, during our quest to become compatible with Mastodon API 4.0.0 (and further cleaning up the codebase in general), and only few of them later, as we find bugs or the Mastodon API changes in incompatible ways.

In order to not bump our major version _too_ often, maybe we can find a way to consolidate at least a good amount of these changes for a `3.0.0` release, and from there continue as suggested?
